src/Sharpliner/AzureDevOps/PublicDefinitions.cs

```diff
@@ -71,6 +80,161 @@ public abstract class JobTemplateDefinition : TemplateDefinition<JobBase>
+            }
+            else if (property.PropertyType == typeof(Step))
+            {
+                parameter = new StepParameter(name, defaultValue: defaultValue as Step);
+            }
+            else if (property.PropertyType == typeof(ConditionedList<Step>))
+            {
+                parameter = new StepListParameter(name, defaultValue: defaultValue as ConditionedList<Step>);
+            }
+            else if (property.PropertyType.IsAssignableFrom(typeof(JobBase)))
+            {
+                parameter = new JobParameter(name, defaultValue: defaultValue as JobBase);
+            }
+            else if (property.PropertyType == typeof(ConditionedList<JobBase>))
+            {
+                parameter = new JobListParameter(name, defaultValue: defaultValue as ConditionedList<JobBase>);
+            }
+            else if (property.PropertyType == typeof(DeploymentJob))
+            {
+                parameter = new DeploymentParameter(name, defaultValue: defaultValue as DeploymentJob);
+            }
+            else if (property.PropertyType == typeof(ConditionedList<DeploymentJob>))
+            {
+                parameter = new DeploymentListParameter(name, defaultValue: defaultValue as ConditionedList<DeploymentJob>);
+            }
+            else if (property.PropertyType == typeof(Stage))
+            {
+                parameter = new StageParameter(name, defaultValue: defaultValue as Stage);
+            }
+            else if (property.PropertyType == typeof(ConditionedList<Stage>))
+            {
+                parameter = new StageListParameter(name, defaultValue: defaultValue as ConditionedList<Stage>);
+            }
+            else
+            {
+                // TODO: use reflection to create an instance of ObjectParameter<T>
+                parameter = new ObjectParameter(name);
+            }
+
+            result.Add(parameter);
+        }
+
+        return result;
+    }
+
+    public TemplateParameters ToTemplateParameters()
+    {
+        var result = new TemplateParameters();
+        var defaultParameters = new TParameters();
+
+        foreach (var property in typeof(TParameters).GetProperties())
+        {
+            var value = property.GetValue(this);
+            if (value is not null && value != property.GetValue(defaultParameters))
+            {
+                var name = property.GetCustomAttribute<YamlMemberAttribute>()?.Alias;
+                name ??= CamelCaseNamingConvention.Instance.Apply(property.Name);
+
+                result.Add(name, value);
+            }
+        }
+
+        return result;
+    }
+}
+
+public abstract class JobTemplateDefinition<TParameters> : JobTemplateDefinition
+    where TParameters : ITemplateParametersProvider, new()
+{
+    public sealed override List<Parameter> Parameters => TemplateParameters.ToParameters();
+
+    [DisallowNull]
+    public TParameters TemplateParameters { get; } = new TParameters();
```

Comment:
Does this need to be public?

Reply:
probably not, this PR is just to add more of the definitions required for `dotnet/xharness` and cleaning up some more API's to see what can work

I expect the good parts of this PR to become small PR's that will be ready to review